repository/src/receive.rs

```diff
@@ -0,0 +1,165 @@
+use super::*;
+use utils::*;
+
+async fn advance_finalized<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    finalized_block_commit_hash: CommitHash,
+    finalization_proof: LastFinalizationProof,
+) -> Result<(), Error> {
+    this.raw.checkout_clean().await?;
+    this.raw
+        .move_branch(FINALIZED_BRANCH_NAME.into(), finalized_block_commit_hash)
+        .await?;
+    this.raw
+        .move_branch(FP_BRANCH_NAME.into(), finalized_block_commit_hash)
+        .await?;
+    this.raw.checkout(FP_BRANCH_NAME.into()).await?;
+    this.raw
+        .create_semantic_commit(format::fp_to_semantic_commit(&finalization_proof))
+        .await?;
+    Ok(())
+}
+
+/// Receives a new branch from a remote peer (either fetch or push).
+///
+/// - Returns `Ok(Ok(()))` if the branch is successfully received.
+/// - Returns `Ok(Err(_))` if the branch is invalid and thus rejected, with the reason.
+/// - Returns `Err(_)` if an error occurs.
+pub async fn receive<T: RawRepository>(
+    this: &mut DistributedRepository<T>,
+    tip_commit_hash: CommitHash,
+) -> Result<Result<(), String>, Error> {
+    let last_finalized_header = this.get_last_finalized_block_header().await?;
+    let last_finalized_commit_hash = this.raw.locate_branch(FINALIZED_BRANCH_NAME.into()).await?;
+    let reserved_state = this.get_reserved_state().await?;
+    let mut csv =
+        CommitSequenceVerifier::new(last_finalized_header.clone(), reserved_state.clone())
+            .map_err(|e| {
+                IntegrityError::new(format!("finalized branch is not accepted by CSV: {e}"))
+            })?;
+
+    if this
+        .raw
+        .find_merge_base(last_finalized_commit_hash, tip_commit_hash)
+        .await?
+        != last_finalized_commit_hash
+    {
+        return Ok(Err(
+            "The commit is not a descendant of the last finalized block.".to_owned(),
+        ));
+    }
+
+    // If the branch ends with a fp commmit
+    if let Ok(last_finalization_proof) =
+        format::fp_from_semantic_commit(this.raw.read_semantic_commit(tip_commit_hash).await?)
+    {
+        // Now we consider the direct parent as the received commit
+        // (fp is not treated in the CSV)
```

Comment:
```suggestion
        // Now we consider the direct parent (the last finalized block commit) as the received commit
        // since finalization proof commits are not treated in the CSV
```

Reply:
No, it's not. `commit_hash` becomes the direct parent of the received commit, which could not be a block commit.

Reply:
It says 
> This branch has only one empty commit that is directly on top of the corresponding block commit, titled with `fp: <height>`.

in the Simperby: the Git-based Blockchain document. Does the documentation have to be fixed?

Reply:
It says 
> This branch has only one empty commit that is directly on top of the corresponding block commit, titled with `fp: <height>`.

in the Simperby: the Git-based Blockchain document. Does the documentation have to be fixed?

Reply:
It could be an invalid branch.

Reply:
Discussed in Discord that my suggestion is only bound to valid cases. In invalid cases, fp branches can be on top of non-block commits.